### PR TITLE
8280357: user.home = "?" when running with systemd DynamicUser=true

### PR DESCRIPTION
**Pull request by RogerRiggs**
In some Linux configurations, the Linux home directory provided by getpwent is not usable.
The value of the system property `user.home` should fallback to the value of $HOME 
if getpwent.user_home is null or less that 2 characters long. "/" is not a valid home directory name.

If $HOME is undefined or empty, the value of the getpwent.user_home is retained.

There are more details in the Jira issue: https://bugs.openjdk.java.net/browse/JDK-8280357

The fix has been tested manually on Ubuntu 20.0.4 using the suggested systemd command line and variations.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8280357](https://bugs.openjdk.java.net/browse/JDK-8280357): user.home = "?" when running with systemd DynamicUser=true
 * [JDK-8282102](https://bugs.openjdk.java.net/browse/JDK-8282102): If the users home directory is invalid, system property user.home is set to $HOME (**CSR**)

### Reviewers
 * [Naoto Sato](https://openjdk.java.net/census#naoto) (@naotoj - **Reviewer**)
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7534/head:pull/7534` \
`$ git checkout pull/7534`

Update a local copy of the PR: \
`$ git checkout pull/7534` \
`$ git pull https://git.openjdk.java.net/jdk pull/7534/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7534`

View PR using the GUI difftool: \
`$ git pr show -t 7534`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7534.diff">https://git.openjdk.java.net/jdk/pull/7534.diff</a>

</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
There is an argument that the JDK should read $HOME first but changing it after 20+ years could be risky, probably difficult to get a list of configurations/environments where the two might yield different locations. So I think the approach is reasonable. Even if "/" were a configured as the home directory in the user entry then $HOME would need to agree so that I think the <2 check is okay too.

**Comment by AlanBateman** [comment]
> I'm uncertain whether the fallback should only be done on Linux to cover the `systemd` case and Docker.
> The need for a fallback seems less applicable on macOs, but since $HOME is usually set to the same value, may be harmless.

I think what you have is okay because it is only used when getpwent doesn't return something useful. It would add complexity if you tried to limit it systemd or containers. I can't think of scenarios on macOS that might trigger the fallback, but it's not a bad fallback to have there too.

**Comment by RogerRiggs** [comment]
The CSR has been approved, please review the PR and the Release note: https://bugs.openjdk.java.net/browse/JDK-8282101

File: src/java.base/unix/native/libjava/java_props_md.c [hunk 1]
```diff
@@ -488,8 +488,14 @@ GetJavaProperties(JNIEnv *env)
...
-        if (sprops.user_home == NULL) {
-            sprops.user_home = "?";
+        if (sprops.user_home == NULL || sprops.user_home[0] == '\0' ||
+            sprops.user_home[1] == '\0') {
+            // If the OS supplied home directory is not defined or less than two characters long
+            // $HOME is the backup source for the home directory, if defined
+            char* user_home = getenv("HOME");
+            if ((user_home != NULL) && (user_home[0] != '\0')) {
+                sprops.user_home = user_home;
+            }
```
Comment (naotoj) [inline src/java.base/unix/native/libjava/java_props_md.c:498]: Is there any possibility where `user.home` is not initialized, and later causes SEGV or NPE? I just wonder the previous version always init to `?` which is odd, but guaranteed not to cause those errors.

Comment (RogerRiggs) [reply src/java.base/unix/native/libjava/java_props_md.c:498]: I can't imagine it not being set. But it is easier to track down the source of a "?" than the source of null.
I thought of changing from "?" to "UNKNOWN" or "NOHOMEDIR" or something but it seems quite remote
after adding the fallback to $HOME.

Comment (naotoj) [reply src/java.base/unix/native/libjava/java_props_md.c:500]: Another option is an empty string `""`, but I don't have a strong preference if it won't throw any errors.

**Review by naotoj — APPROVED** [review]

**Review by AlanBateman — APPROVED** [review]

**Pull request closed without merging**